### PR TITLE
Support iOS devices reporting pressure data of 0

### LINKED ISSUE #28447: Tapping on TextField throws an exception
After the last update of Flutter, TextField has started to behave oddly.  When it is tapped, the following error appears:

```
flutter: ══╡ EXCEPTION CAUGHT BY GESTURE LIBRARY ╞═══════════════════════════════════════════════════════════
flutter: The following assertion was thrown while dispatching a pointer event:
flutter: 'package:flutter/src/gestures/force_press.dart': Failed assertion: line 209 pos 12:
flutter: 'event.pressureMax >= 1.0': is not true.
flutter:
flutter: Either the assertion indicates an error in the framework itself, or we should provide substantially
flutter: more information in this error message to help you determine and fix the underlying cause.
flutter: In either case, please report this assertion by filing a bug on GitHub:
flutter:   https://github.com/flutter/flutter/issues/new?template=BUG.md
flutter:
flutter: When the exception was thrown, this was the stack:
flutter: #2      ForcePressGestureRecognizer.addPointer (package:flutter/src/gestures/force_press.dart:209:12)
flutter: #3      RawGestureDetectorState._handlePointerDown (package:flutter/src/widgets/gesture_detector.dart:731:18)
flutter: #4      RenderPointerListener.handleEvent (package:flutter/src/rendering/proxy_box.dart:2643:14)
flutter: #5      _WidgetsFlutterBinding&BindingBase&GestureBinding.dispatchEvent (package:flutter/src/gestures/binding.dart:192:22)
flutter: #6      _WidgetsFlutterBinding&BindingBase&GestureBinding._handlePointerEvent (package:flutter/src/gestures/binding.dart:149:7)
flutter: #7      _WidgetsFlutterBinding&BindingBase&GestureBinding._flushPointerEventQueue (package:flutter/src/gestures/binding.dart:101:7)
flutter: #8      _WidgetsFlutterBinding&BindingBase&GestureBinding._handlePointerDataPacket (package:flutter/src/gestures/binding.dart:85:7)
flutter: #12     _invoke1 (dart:ui/hooks.dart:223:10)
flutter: #13     _dispatchPointerDataPacket (dart:ui/hooks.dart:144:5)
flutter: (elided 5 frames from class _AssertionError and package dart:async)
flutter:
flutter: Event:
flutter:   PointerDownEvent(Offset(148.3, 711.3))
flutter: Target:
flutter:   RenderPointerListener#a4a39 relayoutBoundary=up15
flutter: ════════════════════════════════════════════════════════════════════════════════════════════════════
```

## Logs

```
[✓] Flutter (Channel master, v1.2.3-pre.31, on Mac OS X 10.14.3 18D109, locale en-US)
    • Flutter version 1.2.3-pre.31 at /Users/user/Android/flutter
    • Framework revision 3423c831b9 (9 hours ago), 2019-02-24 21:48:50 +0100
    • Engine revision 0546e1e5e9
    • Dart version 2.1.2 (build 2.1.2-dev.0.0 c92d5ca288)

[✓] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
    • Android SDK at /Users/user/Android/sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-28, build-tools 28.0.3
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1248-b01)
    • All Android licenses accepted.

[✓] iOS toolchain - develop for iOS devices (Xcode 10.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 10.1, Build version 10B61
    • ios-deploy 1.9.4
    • CocoaPods version 1.5.3

[✓] Android Studio (version 3.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 33.1.1
    • Dart plugin version 182.5215
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1248-b01)

[✓] VS Code (version 1.31.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 2.23.0

[✓] Connected device (1 available)
    • iPhone 6 Plus • 0CBD3FC1-2E7E-4BD2-8728-089A51AD40CA • ios • iOS 12.1 (simulator)

• No issues found!
```

### PR DESCRIPTION
**Pull request by xster**
## Description

iOS devices can actually report pressure data of 0. 

## Related Issues

Fixes #28447 

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes tests for *all* changed/updated/fixed behaviors (See [Test Coverage]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]).
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/test/cupertino/text_field_test.dart [hunk 1]
```diff
@@ -1607,44 +1607,83 @@ void main() {
...
+    const int pointerValue = 1;
+    final TestGesture gesture = await tester.createGesture();
+    await gesture.downWithCustomEvent(
+      textfieldStart + const Offset(150.0, 5.0),
+      PointerDownEvent(
+        pointer: pointerValue,
+        position: textfieldStart + const Offset(150.0, 5.0),
+        // iPhone 6 and below report 0 across the board.
+        pressure: 0,
+        pressureMax: 0,
```
Comment (gspencergoog) [inline packages/flutter/test/cupertino/text_field_test.dart]: Wouldn't you also want to test values like 0.5 and -1.0?

File: packages/flutter/test/material/text_field_test.dart [hunk 2]
```diff
@@ -4632,9 +4630,54 @@ void main() {
     );
 
     await gesture.up();
-    await tester.pumpAndSettle();
+    await tester.pump(const Duration(milliseconds: 500));
```
Comment (HansMuller) [inline packages/flutter/test/material/text_field_test.dart]: What's the rationale for these changes? What happens in 500ms that doesn't happen upon `pumpAndSettle()`?

Comment (xster) [reply packages/flutter/test/material/text_field_test.dart]: If it takes 600ms, it's aesthetically incorrect and a bug.

Comment (xster) [reply packages/flutter/test/material/text_field_test.dart]: Also avoiding pumpAndSettle in general https://docs.flutter.io/flutter/flutter_test/WidgetTester/pumpAndSettle.html

Comment (HansMuller) [reply packages/flutter/test/material/text_field_test.dart]: I don't think it makes sense to avoid replace pumpAndSettle() with arbitrary time limits. Where does 500ms come from?

Comment (jslavitz) [reply packages/flutter/test/material/text_field_test.dart]: We definitely need to figure out a guideline for pump v pumpAndSettle because it seems there are a lot of different opinions on what is best.

Comment (xster) [reply packages/flutter/test/material/text_field_test.dart]: Actually toolbars aren't animated. Replaced all with just a one frame pump. 

I strongly advocate making an effort to bring the usages of pumpAndSettle to 0 in our tests. Animations that doesn't take exactly the amount of time as expected is a bug. UI events that takes more post frame callbacks than exactly the number of expected is a bug.

File: packages/flutter/test/cupertino/text_field_test.dart (hunk 1 above)
Comment (xster) [reply packages/flutter/test/cupertino/text_field_test.dart]: Good point. Though not in this class. Added to force_press_test.dart.

File: packages/flutter/test/material/text_field_test.dart (hunk 2 above)
Comment (HansMuller) [reply packages/flutter/test/material/text_field_test.dart]: I don't think that's a good idea. If you don't have a concrete rationale for a particular time limit then you've added specificity to a test that doesn't belong and will only serve to confuse developers in the future. Many of the time limits in the existing (including this one apparently) are not based on any documented rationale, they're effectively arbitrary.

Comment (xster) [reply packages/flutter/test/material/text_field_test.dart]: I think tests should test production behavior. If a test doesn't have concrete rationale for the time duration of a UI behavior or if it's arbitrary, it should imply that the production behavior has no rationale for being bound by any time duration or is arbitrary. 

That should only happen when the production behavior is intentionally arbitrary (in which case using pumpAndSettle makes sense). If the production value is not arbitrary, the test should not be arbitrary. In this case, it's a single frame.

Comment (HansMuller) [reply packages/flutter/test/material/text_field_test.dart]: Sure, but you can't just make up "production values", they need to be based on real expectations about behavior. And if you're going to include such values then you need to explain what they're based on. Otherwise you're just adding noise to the test implementation.

Comment (xster) [reply packages/flutter/test/material/text_field_test.dart]: If you meant all pumps with a specific duration should have an accompanying code comment, I agree.

Comment (xster) [reply packages/flutter/test/material/text_field_test.dart]: i.e. the way I had it originally is incorrect

Comment (HansMuller) [reply packages/flutter/test/material/text_field_test.dart]: It can take a considerable amount of analysis to decide exactly how much time and/or how many frames a change needs before the state of the UI can be verified.  When a test author is willing to do that and to explain it in a comment, then I agree that explicit pumps and durations are preferred. When there isn't time to do any of that, I think using pumpAndSettle() is preferable, since it indicates an opportunity to tighten up the test.

File: packages/flutter/test/cupertino/text_field_test.dart [hunk 3]
```diff
@@ -1607,44 +1607,83 @@ void main() {
...
+    );
+    // We expect the force press to select a word at the given location.
+    expect(
+      controller.selection,
+      const TextSelection(baseOffset: 8, extentOffset: 12),
+    );
+
+    await gesture.up();
+    await tester.pump();
+    expect(find.byType(CupertinoButton), findsNWidgets(3));
```
Comment (HansMuller) [inline packages/flutter/test/cupertino/text_field_test.dart]: I think what we're doing here is verifying that the cut/copy/paste menu is up?

File: packages/flutter/test/material/text_field_test.dart [hunk 4]
```diff
@@ -4632,9 +4630,54 @@ void main() {
...
+            controller: controller,
+          ),
+        ),
+      ),
+    );
+
+    final Offset textfieldStart = tester.getTopLeft(find.byType(TextField));
+
+    const int pointerValue = 1;
+    final Offset offset = textfieldStart + const Offset(150.0, 5.0);
```
Comment (HansMuller) [inline packages/flutter/test/material/text_field_test.dart]: Here and elsewhere: is this offset suppose to match a location in the controller's text?

**Review by HansMuller — APPROVED** [review]
LGTM

File: packages/flutter/test/cupertino/text_field_test.dart (hunk 3 above)
Comment (xster) [reply packages/flutter/test/cupertino/text_field_test.dart]: Yes. Most places had comments but some were missing. Added.

File: packages/flutter/test/material/text_field_test.dart (hunk 4 above)
Comment (xster) [reply packages/flutter/test/material/text_field_test.dart]: Yes. Do you prefer using `textOffsetToPosition`? I don't have a strong opinion though I felt textOffsetToPosition was a bit round-about-ish.

Comment (HansMuller) [reply packages/flutter/test/material/text_field_test.dart]: It's no big deal, I was just wondering what we expected the x=150 to line up with.
